extensions/amp-next-page/0.1/next-page-service.js

```diff
@@ -497,19 +515,59 @@ export class NextPageService {
 
   /**
    * Sets the specified document as active, updating the document title and URL.
-   * @param {!DocumentRef} documentRef Reference to the document to set as
-   *     active.
+   *
+   * @param {number} documentIndex Index of the document to be activated
    * @private
    */
-  setActiveDocument_(documentRef) {
-    const {amp} = documentRef;
-    this.win_.document.title = amp.title || '';
-    this.activeDocumentRef_ = documentRef;
-    this.setActiveDocumentInHistory_(documentRef);
+  setActiveDocument_(documentIndex) {
+    this.documentRefs_.forEach((docRef, index) => {
+      const {amp} = docRef;
+      let updatedVisibilityState;
+      // Update the title and history
+      if (index == documentIndex) {
+        this.win_.document.title = amp.title || '';
+        this.activeDocumentRef_ = docRef;
+        this.setActiveDocumentInHistory_(docRef);
+        updatedVisibilityState = VisibilityState.VISIBLE;
+      } else if (index !== 0) {
+        updatedVisibilityState = VisibilityState.HIDDEN;
+      }
+      // Show the active doc and hide other docs
+      if (updatedVisibilityState) {
+        this.setDocumentVisibility_(index, updatedVisibilityState);
+      }
+    });
 
     // TODO(emarchiori): Consider updating position fixed elements.
   }
 
+  /**
+   * Manually overrides the document's visible state to the given state
+   *
+   * @param {number} documentIndex Index of the document to change
+   * @param {!../../../src/visibility-state.VisibilityState} visibilityState
+   * @private
+   */
+  setDocumentVisibility_(documentIndex, visibilityState) {
+    // Prevent updating visibility of the host document
+    if (documentIndex === 0) {
+      return;
+    }
+
+    const doc = this.documentRefs_[documentIndex];
+
+    if (doc && doc.amp && doc.amp['ampdoc']) {
```

Comment:
Rather than nesting, I think it would be cleaner to make these early bailouts as well:

```javascript
const ampDoc = doc && doc.amp && doc.amp['ampdoc'];
if (!ampDoc) {
  return;
}

if (!ampDoc.hasBeenVisible() && visibilityState == VisibilityState.HIDDEN) {
  return;
}
```

Also, the access of `ampdoc` here is inconsistent with below, which does not use quotes. Since the type of `amp` is `!Object`, I think the latter (property access) is correct.